include/osquery/registry.h

```diff
@@ -15,14 +15,294 @@
+    if (instance().registries_.count(registry_name) > 0) {
+      return instance().registries_[registry_name]->call(
+          item_name, request, response);
+    }
+
+    return Status(1, "Cannot call " + registry_name + ":" + item_name);
+  }
+
+  static size_t count() { return instance().registries_.size(); }
+
+  static size_t count(const std::string& registry_name) {
+    return instance().registry(registry_name)->count();
+  }
+
+ protected:
+  RegistryFactory() {}
+
+ public:
+  std::map<std::string, RegistryTypeAPIRef> registries_;
+};
+
+class Plugin {
+ public:
+  /// The plugin may perform some initialization, not required.
+  virtual Status setUp() { return Status(0, "Not used"); }
+  /// The plugin may perform some tear down, release, not required.
+  virtual void tearDown() {}
+  /// The plugin may publish route info (other than registry type and name).
+  virtual RouteInfo routeInfo() {
+    return {};
+  }
+  /// The plugin will act on a serialized request, and if a response is needed
+  /// (response is set to true) then response should be a reference to a
+  /// string ready for a serialized response.
+  virtual Status call(const PluginRequest& request, PluginReponse& response) {
+    return Status(1, "Error");
+  }
+
+  /// Allow the plugin to introspect into the registered name (for logging).
+  void setName(const std::string& name) { name_ = name; }
+
+ protected:
+  std::string name_;
+};
+
+template <class... FuncArgs>
+class RegistryTemplate : private boost::noncopyable {
+ public:
+  typedef std::function<void(FuncArgs...)> Func;
+  static const int kDefaultPriority = 100000;
+
+  RegistryTemplate() : alreadyRan_(false) {}
+
+  /**
+   * Registers a function to be invoked when 'run()' is called; fails
+   * if run() has already been called.  Functions will be run with lowest
+   * priority first, with FIFO order on ties.
+   *
+   * Function isn't allowed to throw, calling registerFunc() from within
+   * registered function will cause deadlock.
+   */
+  bool registerFunc(Func func, int priority = kDefaultPriority) {
+    std::lock_guard<std::mutex> guard(mutex_);
+    if (alreadyRan_) {
+      return false;
+    }
+
+    funcMap_[priority].push_back(std::move(func));
+    return true;
+  }
+
+  /**
+   * Runs all functions already registered with registerFunc(), in
+   * order from lowest priority to highest (undefined order on ties).
+   */
+  bool run(FuncArgs... args) noexcept {
+    std::lock_guard<std::mutex> guard(mutex_);
+    if (alreadyRan_) {
+      return false;
+    }
+
+    for (const auto& kv : funcMap_) {
+      for (const auto& func : kv.second) {
+        if (func) {
+          func(args...);
+        }
+      }
+    }
+
+    alreadyRan_ = true;
+    return true;
+  }
+
+ private:
+  std::mutex mutex_;
+  bool alreadyRan_;
+
+  std::map<int, std::vector<Func> > funcMap_;
+};
+
+// NOTE: T should have public or protected default ctor.
+template <class T>
+class Singleton : private T {
```

Comment:
Can we inherit from this singleton class in the above classes that use singletons?